File: android/app/src/main/java/com/reactredispubsub/RedisHandler.java
package com.reactredispubsub;

import android.util.Log;

import java.util.logging.Handler;

import com.lambdaworks.redis.RedisClient;
import com.lambdaworks.redis.pubsub.RedisPubSubAdapter;
import com.lambdaworks.redis.pubsub.RedisPubSubConnection;

import com.reactredispubsub.ServiceBridge;

/**
 * A handler to a Redis PubSub client. Can handle multiple subscribed channels at a time.
 */
public class RedisHandler {

    private String redisURI;    // Redis connection  URI: redis://[password@]host[:port][/databaseNumber]
    private RedisPubSubConnection<String, String> redisInConnection;  // the Redis connection for subscribe
    private RedisPubSubConnection<String, String> redisOutConnection;  // the Redis connection for publish
    private RedisPubSubAdapter<String, String> listener;    // the Redis channels listener

    private ServiceBridge bridgeModule;

    private static RedisHandler singletonHandler;   // singleton instance

    /**
     * Get the singleton Redis handler instance
     * 
     * @param redisURI the Redis connection URI
     * @param module the ServiceBridge instance
     * 
     * @return the handler singleton instance
     */
    public static RedisHandler getInstance(String redisURI, ServiceBridge module) {
        if(singletonHandler == null) {
            singletonHandler = new RedisHandler(redisURI, module);
        }
        return singletonHandler;
    }

    /**
     * Create a new Redis connection client with a single listener for all channels subscribed
     * 
     * @param redisURI the Redis connection URI
     * @param module the ServiceBridge instance
     */
    private RedisHandler(String redisURI, ServiceBridge module) {
        this.redisURI = redisURI;
        this.bridgeModule = module;

        // Create connection
        RedisClient client = RedisClient.create(this.redisURI);
        redisInConnection = client.connectPubSub();
        redisOutConnection = client.connectPubSub();

        // Create and add the listener
        listener = new RedisPubSubAdapter<String, String>() {
            @Override
            public void message(String channel, String message) {
                singletonHandler.bridgeModule.handleMessage(channel, message);
            }
        };
        redisInConnection.addListener(listener);
    }

    /**
     * Subscribe a PubSub channel
     * 
     * @param channel the redis channel name
     */
    public void subscribePubSub(String channel) {
        redisInConnection.subscribe(channel);
    }

    /**
     * Unsubscribe a PubSub channel
     * 
     * @param channel the Redis channel name
     */
    public void unsubscribePubSub(String channel){
        if(redisInConnection != null) {
            redisInConnection.unsubscribe(channel);
        }
    }

    /**
     * Publish message on a PubSub channel
     * 
     * @param channel the Redis channel name
     * @param message the message to publish
     */
    public void publishPubSub(String channel, String message) {
        if(redisOutConnection != null) {
            redisOutConnection.publish(channel, message);
        }
    }
}

File: ios/RedisHandler.swift
//
//  RedisHandler.swift
//  UberServicosProvider
//
//  Created by Codificar Sistemas on 09/08/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import PSSRedisClient

class RedisHandler: NSObject, RedisManagerDelegate {
  
  enum RedisHandlerError: Error {
    case clientNotConnected
  }
  
  private var redisHost: String?
  private var redisPort: Int?
  private var redisPwd: String?
  
  private var redisInConnection: RedisClient?  // the Redis connection for subscribe
  private var redisOutConnection: RedisClient?  // the Redis connection for publish
  
  private var bridgeModule: ServiceBridge?
  
  private static var singletonHandler: RedisHandler? // singleton instance
  private var canSub: Bool // control if client is already connected to subscribe
  
  /**
   * Get the singleton Redis handler instance
   *
   * @param redisURI the Redis connection URI
   * @param module the BubbleServiceBridgeModule instance
   *
   * @return the handler singleton instance
   */
  static func getInstance(redisURI: String, module: ServiceBridge) throws -> RedisHandler {
    if(RedisHandler.singletonHandler == nil) {
      RedisHandler.singletonHandler = try RedisHandler(redisURI: redisURI, module: module)
    }
    return RedisHandler.singletonHandler!
  }
  
  /**
   * Create a new Redis connection client with a single listener for all channels subscribed
   *
   * @param redisURI the Redis connection URI - redis://[password@]host[:port][/databaseNumber]
   * @param module the BubbleServiceBridgeModule instance
   */
  private init(redisURI: String?, module: ServiceBridge) throws {
    
    self.canSub = false;
    
    super.init()
    
    self.bridgeModule = module
    
    // Get Redis URI params
    let index = redisURI!.index(redisURI!.startIndex, offsetBy: 8)
    let uri = String(redisURI!.suffix(from: index))
    let params = uri.replacingOccurrences(of: "@", with: ":").components(separatedBy: ":")
    
    self.redisPwd = params[0]
    self.redisHost = params[1]
    self.redisPort = Int(params[2])
    
    // Create redis connections
    self.redisInConnection = RedisClient(delegate: self)
    self.redisOutConnection = RedisClient(delegate: self)
    self.redisInConnection?.connect(host: self.redisHost!, port: self.redisPort!, pwd: self.redisPwd!)
    self.redisOutConnection?.connect(host: self.redisHost!, port: self.redisPort!, pwd: self.redisPwd!)
  }
  
  /**
   * Delegate the message received via subscribed channel to
   * the bridge module
   * @param channel the redis channel subscribed
   * @param message the message received
   */
  func subscriptionMessageReceived(channel: String, message: String) {
    debugPrint("REDIS: SUB message received")
    
    self.bridgeModule?.handleMessage(channel: channel, message: message)
  }
  
  func socketDidDisconnect(client: RedisClient, error: Error?) {
    debugPrint("REDIS: Disconnected (Error: \(String(describing: error?.localizedDescription)))")
    
    // Try to reconnect the client
    client.connect(host: self.redisHost!, port: self.redisPort!, pwd: self.redisPwd!)
  }
  
  func socketDidConnect(client: RedisClient) {
    debugPrint("REDIS: Socket connected");
    
    self.canSub = self.redisInConnection?.isConnected() ?? false
  }
  
  func socketDidSubscribe(socket: RedisClient, channel: String) {
    debugPrint("REDIS: Subscribed to \(channel)")
  }
  
  func socketDidReceivePong(socket: RedisClient) {
    //    debugPrint("REDIS: Pong received")
  }
  
  /**
   * Subscribe to a PubSub channel
   * (5 attempts, 1 second between each attempt)
   * @param channel the redis channel name
   */
  func subscribePubSub(channel: String, attempt: Int = 0) throws{
    if(self.canSub) {
      self.redisInConnection?.subscribe(to: channel)
    }
    else if(attempt < 5){
      debugPrint("REDIS: Waiting for socket connection to subscribe...")
      /* ENSURE this script is NOT running on the main thread, otherwise the sleep
       command will freeze the application. See: requiresMainQueueSetup */
      sleep(1)
      try self.subscribePubSub(channel: channel, attempt: attempt+1)
    }
    else {
      throw RedisHandlerError.clientNotConnected
    }
  }
  
  /**
   * Unsubscribe a PubSub channel
   *
   * @param channel the Redis channel name
   */
  func unsubscribePubSub(channel: String) throws {
    self.redisInConnection?.exec(args: ["UNSUBSCRIBE", channel], completion: self.messageReceived)
    debugPrint("REDIS: successfully unsubscribed from channel \(channel)")
  }
  
  func messageReceived(message: NSArray) {
    debugPrint("REDIS: return message received")
    if (message.firstObject as? NSError != nil) {
      let error = message.firstObject as! NSError
      let userInfo = error.userInfo
      
      if let possibleMessage = userInfo["message"] {
        if let actualMessage = possibleMessage as? String {
          debugPrint("REDIS: \(actualMessage)")
        }
      }
    } else {
      debugPrint("REDIS: Results: \(message.componentsJoined(by: " "))")
    }
  }
  
  /**
   * Publish message on a PubSub channel
   *
   * @param channel the Redis channel name
   * @param message the message to publish
   */
  func publishPubSub(_ channel: String, message: String) {
    debugPrint("DEBUG: publish on \(channel) message: \(message)")
    
    self.redisOutConnection?.exec(args: ["PUBLISH", channel, message], completion: self.messageReceived)
  }
  
}
